File: workspace/solutions/ecmascript/src/s0/s2/s3/s9/s_0239_00.ts
/*
 * @lc app=leetcode.cn id=239 lang=typescript
 *
 * [239] 滑动窗口最大值
 */

// @lc code=start
function maxSlidingWindow(nums: number[], k: number): number[] {
    const queue: number[] = [];
    const result: number[] = [];

    for (let i = 0; i < nums.length; i++) {
        const num = nums[i]!;
        if (queue.length > 0 && (i + 1 - k) > queue.at(0)!) {
            queue.shift();
        }
        while (queue.length > 0 && nums[queue.at(-1)!]! <= num) {
            queue.pop();
        }
        queue.push(i);
        if ((i + 1) >= k) {
            result.push(nums[queue.at(0)!]!);
        }
    }
    return result;
}
// @lc code=end

export default maxSlidingWindow;

File: workspace/solutions/kotlin/lib/src/main/kotlin/s0/s2/s3/s9/s_0239_00/s_0239_00.kt
/*
 * @lc app=leetcode.cn id=239 lang=kotlin
 *
 * [239] 滑动窗口最大值
 */

package s0.s2.s3.s9.s_0239_00

// @lc code=start
class Solution {
    fun maxSlidingWindow(nums: IntArray, k: Int): IntArray {
        var queue = mutableListOf<Int>()
        var result = mutableListOf<Int>()

        nums.forEachIndexed { i, num ->
            if (queue.isNotEmpty() && (i + 1 - k) > queue.first()) {
                queue.removeFirst()
            }
            while (queue.isNotEmpty() && nums[queue.last()] <= num) {
                queue.removeLast()
            }
            queue.addLast(i)
            if ((i + 1) >= k) {
                result.add(nums[queue.first()])
            }
        }

        return result.toIntArray()
    }
}
// @lc code=end
